[ENT-1670] 1640822: [RFE] Change or switch content view from content host

wolfdale opened the pull request:
- Allowing updating consumer environment by env name. Modified consumer PUT update API.
- Added spec test for different scenarios.

Ceiu commented on server/src/main/java/org/candlepin/resource/ConsumerResource.java:
```diff
@@ -1397,6 +1377,63 @@ public boolean performConsumerUpdates(ConsumerDTO updated, Consumer toUpdate,
+            }
+            else {
+                throw new NotFoundException(i18n.tr(
+                        "Environment with name \"{0}\" could not be found.", environmentName));
```
Indentation is too long here. If checkstyle is complaining, drop the i18n.tr down a line with it. Same with the identical line below.

wolfdale replied:
No checkstyle complaints.

Ceiu replied:
I meant if checkstyle complains about it not being indented enough. The indentation check is pretty spotty at times.

Ceiu commented on server/src/main/java/org/candlepin/resource/ConsumerResource.java:
```diff
@@ -1397,6 +1377,63 @@ public boolean performConsumerUpdates(ConsumerDTO updated, Consumer toUpdate,
+            List<Environment> environment = environmentCurator.listForOwnerByName(toUpdate.getOwner(),
+                environmentName).list();
+
+            if (!environment.isEmpty() && environment.size() == 1) {
```
This is a redundant check. If the environment collection size is 1, we know it's not empty. Since we're not differentiating between empty and more-than-one in the error output, we can simplify this down to just the size check.

wolfdale replied:
yes sure.

Ceiu commented on server/src/main/java/org/candlepin/resource/ConsumerResource.java:
```diff
@@ -1397,6 +1377,63 @@ public boolean performConsumerUpdates(ConsumerDTO updated, Consumer toUpdate,
+        if (environmentId != null && (toUpdate.getEnvironmentId() == null ||
+            !toUpdate.getEnvironmentId().equals(environmentId))) {
+
+            if (!environmentFound) {
```
environmentFound will never be true at this point, so this check doesn't do anything. If the intent is only to do this lookup if the environment has not yet been looked up (i.e. we already did that lookup by name), then this whole section needs a refactor around that idea.

Ceiu commented on server/src/main/java/org/candlepin/resource/ConsumerResource.java:
```diff
@@ -1397,6 +1377,63 @@ public boolean performConsumerUpdates(ConsumerDTO updated, Consumer toUpdate,
+            log.info("Updating environment to: {}", environmentId);
+        }
+
+        if (e != null && changesMade) {
```
I don't quite follow the intent behind changesMade here. It's set only in one of the lookup cases above, but it's not associated with any actual change. This seems like something that should be set only if an environment was found; but since we throw an exception in both lookup cases, this only serves the purpose of indicating if (a) no environment reference was provided or (b) a reference was provided *and* a change was made as a result. Following that, this should only be set in the block it's currently guarding.

Looking at the original code, that seems to be what it was doing before.

wolfdale replied:
I removed `e != null` check. It seems reasonable that we update only when changesMade flag is true.

Ceiu commented on server/src/main/java/org/candlepin/resource/ConsumerResource.java:
```diff
@@ -1397,6 +1377,63 @@ public boolean performConsumerUpdates(ConsumerDTO updated, Consumer toUpdate,
+            }
+            // lazily regenerate certs, so the client can still work
+            poolManager.regenerateCertificatesOf(toUpdate, true);
+
```
Move this whitespace up two lines to logically separate the regeneration section and the CAC section.

wolfdale replied:
corrected.

Ceiu commented on server/src/main/java/org/candlepin/resource/ConsumerResource.java:
```diff
@@ -1397,6 +1377,63 @@ public boolean performConsumerUpdates(ConsumerDTO updated, Consumer toUpdate,
+        }
+
+        if (environmentId != null && (toUpdate.getEnvironmentId() == null ||
+            !toUpdate.getEnvironmentId().equals(environmentId))) {
```
We can pair this down a bit to drop an extraneous null check:

```if (environmentId != null && !environmentId.equals(toUpdate.getEnvironmentId())) {```

wolfdale commented:
Earlier the incoming envId is compared to existing envId (if any), if these both ids are different only then we set the env and update CAC & re-generate certs.
What I am trying to do is to save extra DB call & prevent unnecessary updates every time if API keep sending same environment name again & again. This is what I have in my mind.

```
Check if envName is not null && envId is null.
  if YES
    Get environment from DB.
    set environment Id , set environment object
    environmentFound = true (Preventing another DB call, if incoming envId is different from what we already have)

    Go on to next IF condition.

Check if envId is not null (This prevent when both envName and envId are null) && envId is different than what we already have.
  
  if YES
    
    if environmentFound = fasle
      Get ENV from DB
    
    Mark changesMade = true

if ChangesMade = true
  Update CAC and re-gen certs.
```

Ceiu commented:
That's good that you're aware of the cost of doing needless database lookups, but let's step through the intent and code to clean this up a bit:

> Check if envName is not null && envId is null.

Straight away you're establishing a priority of attributes to use to lookup. By checking that envId is null, you're asserting that if it's *not* null, we should use that first, and fallback to name only if the ID isn't present. The priority here is important, as we need to establish some rules for the situation when both are provided. More on this later.

>   if YES
>    Get environment from DB.
>    set environment Id , set environment object
>    environmentFound = true (Preventing another DB call, if incoming envId is different from what we already have)
>
>    Go on to next IF condition.

We do the lookup, but then we store a bunch of redundant information -- even at the pseudocode level. If we have the ID to perform a lookup, there isn't much value in storing it in another variable; especially since it isn't expensive to fetch. Additionally, we don't need to store whether or not we found the environment, since the simple presence of an environment entity provides that naturally.

> Check if envId is not null (This prevent when both envName and envId are null) && envId is different than what we already have.

As hinted above, this is the interesting part here. We check the environment ID again to avoid a lookup, but we also verify that it's actually different to skip doing any work if the environment ID is unchanged. While this is fine, this gets a little tricky since we're bolting on a lookup ahead of it. This is the cornerstone of the method in both core functionality and optimization. However, the optimization may be premature here. More on this later.

>  if YES
>    if environmentFound = fasle
>      Get ENV from DB

Continuation of the resolution step, nothing major here other than using the variable we don't actually need. This part is fine in pseudocode, but in real code we can just use the environment object itself.

>    Mark changesMade = true
>
> if ChangesMade = true
>  Update CAC and re-gen certs.

This bit is a tad confusing. If you follow the code, it's understandable, but the naming convention chosen here is poor. We haven't actually made any changes yet, so by flagging it as such, anyone reading the code needs to understand that this is setting up a promise/contract, and that it's following the usage convention of the intended caller rather than the owning method. That last bit is backward in my opinion, and the caller of a utility method should have very little influence over the conventions applied by the method in question.

---

Alright, so with all of that out of the way, we can compare that to the code and what we actually *need* to do.

First, the bulk of this method surrounds environment resolution and validation. We're spending a great deal of effort here to avoid doing a full lookup if possible. That's good, but the optimizations chosen are starting to mix the operations. Moreover, the pseudocode does not make it clear that we have a specific order in which the attributes are ranked *and* that if both are provided we ignore the name -- even if it would lead to cases where the client has specified an invalid pairing.

Secondly, we perform the actual update. Looking at what we actually do with the bits we looked up, we can see that we don't actually need the whole environment, so the work we're doing is very heavy when all we want is a valid environment ID. Beyond that, the extra work we do is ancillary state management when the environment changes, so everything comes down to an ID comparison.

With all that said, we can rearrange the pseudocode as follows:

```
Check if env ID is present
Yes:
  check if specified ID exists
  throw REST exception if it does not
  
Check if we don't yet have a validated env ID AND env name is present
Yes:
  fetch environment ID by org+name
  throw REST exception if it does not exist

Check if we have a validated environment ID
Yes:
  compare to assigned environment ID
  if different:
    update consumer
    update CAC
    mark certs dirty
    set updated flag to true
  
return updated flag
```

This is very close to the work you had, but it rearranges everything in both priority and logical order without mixing the two during optimization. You'll note that we now require two additional DB operations, but that's not really a big deal -- one of them arguably should have already existed, and the other is just the cost of cleaning this code up.

Translated to code you'll get something along the lines of this:
```
public boolean updateEnvironment(ConsumerDTO update, Consumer toUpdate) {
  if (update.getEnvironment() == null) {
    // No updates being made here, don't bother attempting anything else
    return false;
  }

  boolean changesMade = false;
  String validatedEnvId = null;  
  String envId = update.getEnvironment().getId();
  String envName = update.getEnvironment().getName();
  
  if (envId != null) {
    if (!this.environmentCurator.exists(envId) {
      throw new NotFoundException(...);
    }
    
    validatedEnvId = envId;
  }
  
  if (validatedEnvId == null && envName != null) {
    validatedEnvId = this.environmentCurator.getEnvironmentIdByName(org, envName);
    
    if (validatedEnvId == null) {
      throw new NotFoundException(...);
    }
  }
  
  if (validatedEnvId != null && !validatedEnvId.equals(toUpdate.getEnvironmentId())) {
    ...
    
    changesMade = true;
  }
  
  return changesMade;
}
```

And if you *really* want to optimize out that existence check, you'd modify the first block as follows:

```
if (envId != null) {
  if (!envId.equals(toUpdate.getEnvironmentId()) && !this.environmentCurator.exists(envId) {
    throw new NotFoundException(...);
  }
  
  validatedEnvId = envId;
}
```

From here, the two DB calls would need to be added to AbstractHibernateCurator and EnvironmentCurator, respectively. Both can be done pretty quickly using a one-liner JPQL query.

Ceiu commented:
One thing I forgot to mention that you should be aware of, and why the conflating of the two operations is kind of bad: right now the resource is acting as a resource and controller. As work continues to separate those out, the validation bits (where the REST exceptions originate) will be done in the resource and the actual update bits will occur in the controller. Exactly how we go about that will vary as we hit each resource, but we need to be mindful of that separation more than we have been in the past.

wolfdale commented:
One corner cases is missed here, if the environment is to be updated for first time.
Added another check for it, at line 1395.
```
if (environmentId != null && (toUpdate.getEnvironmentId() == null ||
            !toUpdate.getEnvironmentId().equals(environmentId))) {
```
`This is very close to the work you had, but it rearranges everything in both priority and logical order without mixing the two during optimization. `
I did exact opposite of it, by prioritizing envName, which seems to work but was not cleaner approach. 

`You'll note that we now require two additional DB operations, but that's not really a big deal -- one of them arguably should have already existed, and the other is just the cost of cleaning this code up.
`
Earlier I used older Curator method just for code re-usability to avoid any redundant DB operations.

Also I have added a `exists` method at AbstractHibernateCurator. Notice that its a find operation, was trying with `entitymanager.getReference()` method but it return a proxy object and will not hit DB unit any GET method is called on it. Any suggestion here?

```
    public boolean exists(Serializable id) {
        return this.getEntityManager().find(entityType, id) != null;
    }
```
Also would request you to not post code it defeat the purpose to explore on different viewpoint.

Ceiu commented:
> I did exact opposite of it, by prioritizing envName,
Something important to note, that while you had the name being *processed* first, because the processing was contingent on the ID being null. Thus, while the name was processed first in the code, it had lower logical priority than the ID.

For the exists method, we want to avoid fetching the object if at all possible, since we (a) don't need any of the data and (b) don't want to trigger secondary lookups for nested collections or whatever. Following that, we probably want to steer closer to doing a COUNT-based lookup. We can do this in one of two ways: JPQL or JPA criteria query. The latter is probably safest so we're not doing our own string concatenation, but either is acceptable.

The one nibble here is that at the AHC level, we don't know *exactly* what the ID column name is, so we'll have to fetch it from the entity metadata. Beyond that, this is just a simple expression of the SQL ```SELECT count(id) FROM cp_entity_table WHERE id = :id```.

Ceiu commented on server/src/main/java/org/candlepin/model/AbstractHibernateCurator.java:
```diff
@@ -171,6 +170,49 @@ public E get(Serializable id) {
+     * Returns name of primary key column
+     * @return Returns entity primary key column
+     */
+    public String getPrimaryColumnName() {
```
We should follow Hibernate/JPA or SQL naming conventions here: either ```getIdentifierName``` or ```getPrimaryKeyName```.

wolfdale replied:
done.

wolfdale commented on server/src/main/java/org/candlepin/model/EnvironmentCurator.java:
```diff
@@ -172,4 +172,29 @@ public int deleteEnvironmentsForOwner(Owner owner) {
+                        .getSingleResult();
+            }
+            catch (NoResultException exp) {
+                log.debug("Unable to find environment Id by envName {}", envName);
```
Yes, agreed.

wolfdale commented on server/src/main/java/org/candlepin/model/EnvironmentCurator.java:
```diff
@@ -172,4 +172,29 @@ public int deleteEnvironmentsForOwner(Owner owner) {
+            try {
+                envId = this.getEntityManager().createQuery(jpql, String.class)
+                        .setParameter("owner_id", ownerId).setParameter("env_name", envName)
+                        .getSingleResult();
```
Sure, I think this could be included in checkstyle.

wolfdale commented on server/src/main/java/org/candlepin/resource/ConsumerResource.java:
```diff
@@ -1397,6 +1377,64 @@ public boolean performConsumerUpdates(ConsumerDTO updated, Consumer toUpdate,
+        }
+
+        if (changesMade) {
+            toUpdate.setEnvironment(e);
```
This was the part of old code. But it does make sense to update envId only. Fixed it.

wolfdale commented on server/src/main/java/org/candlepin/model/AbstractHibernateCurator.java:
```diff
@@ -171,6 +170,49 @@ public E get(Serializable id) {
+
+        if (columnName != null) {
+            String jpql = "SELECT count(t) FROM " + entityType.getCanonicalName() +
+                " t WHERE " + columnName + " = :id";
```
Changing to JPA criteria.

wolfdale commented on server/src/main/java/org/candlepin/resource/ConsumerResource.java:
```diff
@@ -1397,6 +1377,64 @@ public boolean performConsumerUpdates(ConsumerDTO updated, Consumer toUpdate,
+        String environmentName = updated.getEnvironment() == null ? null : updated.getEnvironment().getName();
+        String validatedEnvId = null;
+
+        if (environmentId != null && (toUpdate.getEnvironmentId() == null ||
```
This is what we had in first iteration, IF-ELSE logic. its just I prioritized the envName, resulting in close coupling revolving around environment entity. Now that environment object is totally removed, making it make cleaner and easier to make sense.

My primary objective was to adjust environment name without touching old code, being cautious not to break things that I don't have good understanding on.

Ceiu replied:
Sometimes that's fine, but for situations like this where the order fields are processed and the way certain actions are done, it's best to redo the whole block so the input semantics make sense and there's no confusing side effects.

wolfdale commented on server/src/main/java/org/candlepin/model/AbstractHibernateCurator.java:
```diff
@@ -171,6 +170,49 @@ public E get(Serializable id) {
+    public String getPrimaryColumnName() {
+        String primaryColumn = null;
+
+        for (SingularAttribute attribute : this.getEntityManager().
```
Doesn't both approaches require to have entity objects. 
For
1)  We can get identifier from `EntityManagerFactory` this is supported only for JPA 2.0 onward, but it require entity object as parameter.    
```
Object key = this.getEntityManager()
            .getEntityManagerFactory()
            .getPersistenceUnitUtil()
            .getIdentifier(entityObj);
```
2) Hibernate session also require entity object. 
```
org.hibernate.Session session = (Session) this.currentSession().getDelegate();
Serializable key = session.getIdentifier(entity_obj);
```

Ceiu replied:
You're not looking for the value of the identifier field, you're looking for the field name, which doesn't require an instance of the entity.

JPA:
```
String name = entityManager.getMetamodel()
  .entity(<entity class>)
  .getId()
  .getName();
```

Hibernate:
```
String name = ((SessionImpl) session)
  .getSessionFactory()
  .getClassMetadata(<entity class>)
  .getIdentifierPropertyName();
```

wolfdale replied:
Regarding composite keys, JPA `getId()` throws `IllegalArgumentException` exception, tested with `OwnerProduct` entity which has `OwnerProductKey` as key. For now I decided to suppress the exception and return null. 
Also added a test case with composite key in AHC class.

wolfdale commented:
Added more test around AHC & EnvironmentCurator classes.
